operator/controllers/contour_ingress.go

```diff
@@ -0,0 +1,301 @@
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	v12 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	types2 "k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"knative.dev/pkg/kmp"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	ENV_CONTOUR_ENABLED       = "CONTOUR_ENABLED"
+	ENV_CONTOUR_BASE_HOST     = "CONTOUR_BASE_HOST"
+	ENV_CONTOUR_INGRESS_CLASS = "CONTOUR_INGRESS_CLASS"
+)
+
+var grpcProtocol = "h2c"
+
+// +kubebuilder:rbac:groups=projectcontour.io,resources=httpproxies,verbs=get;list;watch;create;update;patch;delete
+
+type ContourIngress struct {
+	client   client.Client
+	recorder record.EventRecorder
+	scheme   *runtime.Scheme
+}
+
+func NewContourIngress() Ingress {
+	return &ContourIngress{}
+}
+
+func (i *ContourIngress) AddToScheme(scheme *runtime.Scheme) {
+	contour.AddKnownTypes(scheme)
+}
+
+func (i *ContourIngress) SetupWithManager(mgr ctrl.Manager) ([]runtime.Object, error) {
+	// Store the client, recorder and scheme for use later
+	i.client = mgr.GetClient()
+	i.recorder = mgr.GetEventRecorderFor(constants.ControllerName)
+	i.scheme = mgr.GetScheme()
+
+	// Index on VirtualService
+	if err := mgr.GetFieldIndexer().IndexField(&contour.HTTPProxy{}, ownerKey, func(rawObj runtime.Object) []string {
+		// grab the deployment object, extract the owner...
+		vsvc := rawObj.(*contour.HTTPProxy)
+		owner := metav1.GetControllerOf(vsvc)
+		if owner == nil {
+			return nil
+		}
+		// ...make sure it's a SeldonDeployment...
+		if owner.APIVersion != apiGVStr || owner.Kind != "SeldonDeployment" {
+			return nil
+		}
+
+		// ...and if so, return it
+		return []string{owner.Name}
+	}); err != nil {
+		return nil, err
+	}
+	return []runtime.Object{&contour.HTTPProxy{}}, nil
+}
+
+func (i *ContourIngress) GeneratePredictorResources(mlDep *v1.SeldonDeployment, seldonId string, namespace string, ports []httpGrpcPorts, httpAllowed bool, grpcAllowed bool) (map[string][]runtime.Object, error) {
+	contourBaseHost := GetEnv(ENV_CONTOUR_BASE_HOST, "") // TODO(jpg): This to be better to handle cases when base host isn't set better
+	contourIngressClass := GetEnv(ENV_CONTOUR_INGRESS_CLASS, "")
+	var annotations map[string]string
+	if contourIngressClass != "" {
+		annotations = map[string]string{
+			"projectcontour.io/ingress.class": contourIngressClass,
+		}
+	}
+	var httpProxies []runtime.Object
+	var routes []contour.Route
+	var grpcRoutes []contour.Route
+
+	for i, predictor := range mlDep.Spec.Predictors {
+		predictorServiceName := v1.GetPredictorKey(mlDep, &predictor)
+		routes = append(routes, contour.Route{
+			Conditions: []contour.Condition{{
+				Prefix: "/",
+			}},
+			Services: []contour.Service{
+				{
+					Name:   predictorServiceName,
+					Mirror: predictor.Shadow,
```

Comment:
we need validate predictor.Shadow, and predictor.Traffic
e.g. the default value is False?
Total traffic should be 100.